### ISSUE TITLE
Enhancement: [no-unnecessary-type-assertion] also flag redundant satisfies expressions

### ISSUE THREAD

**Issue by lo1tuma** [typescript-eslint/typescript-eslint#12402]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have searched for related issues and found none that match my proposal.
- [x] I have searched the current rule list and found no rules that match my proposal.
- [x] I have read the FAQ and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community.

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-type-assertion/

### Description

`no-unnecessary-type-assertion` already flags `as T` and `!` when the expression is already assignable to the asserted type. `satisfies T` has the same shape — a type check that can be a no-op — but the rule never reports it. Today the rule's source only references `TSSatisfiesExpression` internally to skip surrounding contextual-type logic; it does not check the satisfies expression itself.

Cases worth reporting: a `satisfies T` whose expression is already contextually typed as `T` (or stricter) via an enclosing annotation, parameter type, return type, or another `satisfies`/`as` on the outside.

### Fail

```ts
function listen(options: { port: number }) {}
listen({ port: 3000 } satisfies { port: number });

const config: { port: number } = { port: 3000 } satisfies { port: number };

const nested = ({ a: 1 } satisfies { a: number }) satisfies { a: number };

function foo(): Bar {
  return {} satisfies Bar;
}
```

### Pass

```ts
function listen(options: { port: number }) {}
listen({ port: 3000 });

const config: { port: number } = { port: 3000 };

const nested = { a: 1 } satisfies { a: number };

function foo(): Bar {
  return {};
}
```

### Additional Info

Could ship behind an option (e.g. `checkSatisfies: boolean`, default `false`) to avoid churn for users who deliberately keep `satisfies` as documentation. Implementation can mirror the existing `as`-redundancy path: compare the contextual type at the satisfies node against the expression type.

**Comment by kirkwaiblinger** [comment]
This is intriguing! I can see that this pattern is definitely something that can reasonably be linted for, if it's a problem in the real world...

I'm skeptical that this is a problem in the real world, though. `satisfies` aren't super common in my experience (and they're used for pretty specific scenarios when they are used), and it's hard to see redundant/unnecessary `satisfies` having any meaningful harmful effect (by construction, the worst it could do is visual clutter). So, to me the value proposition is low, and for that reason I'm overall -0.5 to this being part of typescript-eslint, I think.

I'm also -0.5 on this check being part of `no-unnecessary-type-assertion` (it isn't a type assertion - it's just a `satisfies`), but it could just be its own rule if we went forward with this, so that's not really a problem.

Would like to hear thoughts from other team members @typescript-eslint/triage-team

**Comment by bradzacher** [comment]
The most common usage of them that I see them is cases where people actively asserting that a type is something so that they have a typecheck error when it's not -- i.e. every `satisfies` I've seen is redundant!! That is -- until it's not and it's caught a bug.

Unlike `as` which actively mutates the underlying type to widen or narrow it -- a `satisfies` expression can either narrow a type or do nothing.

For example `'a' statisfies 'a'` does nothing in most cases.
But cases like this actually **_DO_** do something!
```ts
const x = { a: 1 };
//    ^? { a: number }
const y = { a: 1 } satisfies { a: 1 };
//    ^? { a: 1 }
```
[playground](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAHjAvDA3jAhgLhgRhgXwG4AoAelJkpgD0B+Y0SWATyVQ2z3xgnSgEsIAM34BTCOyy4CJclRr0gA)

This is a surprising effect that most people don't realise from `satisfies`!
Now -- you might say "so what? this doesn't really change anything" and for most cases you're right! But if you slap an `export` on the above line well you can see where this can change things -- now the generated `.d.ts` file is actually different because of the `satisfies`!

The problem with this in particular is that encoding all of the rules where `satisfies` results in a meaningfully different type would mean rewriting TS's algorithms around `satisfies` -- which we're not going to do. We also can't just do "what is the type before and after the `satisfies`?" [because the `satisfies` physically modifies the underlying type object of the "before" to match the type "after"](https://typescript-eslint.io/play/#ts=6.0.2&showAST=types&fileType=.tsx&code=MYewdgzgLgBAHjAvDA3jAhgLhgRhgXwG4YTSzzyB6SmASzADMBTAJxaYBMYoBPAByZIAfKgzYwAVwC2AI1YEAsAChQkWDySisuAjAjootCA1pMIW7HiIxqdRq3ZdeA4RZ35lyoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)-- so to any naive check it looks the same and thus would be flagged as redundant!

It's also important to note that this mutation doesn't work exactly as you might think either -- this mutation does not mean that for `export const a = b satisfies c` either `export const a: b` or `export const a: c` are correct! There are cases like this where `export const a: d`!
```ts
type T = { readonly c: 1 }
const x = { a: 1, b: { c: 1 } };
//    ^? { a: number, b: { c: number } }
const y = { a: 1, b: { c: 1 } } satisfies { a: 1, b: T };
//    ^? { a: 1, b: { c: 1 } }
```
[playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAKlC8UDeUBOECGATA9gOwBsQoBjALigEYoBfAKBPwGdgoAPBZKDCygGigAjCinJVatANx0A9DKgKoAPQD8DZq2KIUPKgOFcx1GhKYZgASyYAzCxCZdd-IRTg1pcxcrVA)
Again you might say that the difference between `{ a: 1, b: { c: 1 } }` and `{ a: 1, b: T }` is no difference at all given that `{ c: 1 }` is assignable to `T` -- but the subtlety is that the inverse is not true and that's a subtle difference that can solve or introduce down stream type errors.

So yeah:
- From my experience the vast, vast, vast majority of `satisfies` are intentionally redundant to guard against bug cases.
   - This automatically makes it fail the "useful to the broader TypeScript community" check
- I would guess that the majority of the minority cases a intentional "mutation" cases as I mentioned
- Then the remaining minority of the minority cases would be truly redundant cases.

I also agree with Kirk's comments.

So overall I give this a **very** strong -1.
With two maintainers against, we shall reject this proposal.

**Comment by lo1tuma** [comment]
Thanks, both of you for the detailed feedback. I’ve learned a lot about `satisfies` behavior I wasn’t aware of. I agree, that this doesn’t make sense as part of `no-unnecessary-type-assertion`. I also think `redundant` might not be the real problematic patttern I want to have linted. Maybe it should be more something like `no-useless-satisfies`, but then again the term `useless` might not be the same for every user.  
My personal rule of thumb is, that I only use `satisfies` with types that are stricter or narrower than the original type and the intention is to not change the original type. So, given what I have learned from @bradzacher I think I would even like to have two things flagged by such a rule (1) `satisfies` usages where it would change the type and (2) `satisfies` usages where the `satisfies` operand type what by the same or broader than the original type. But you are right, I am not sure how other people think about those patterns as being problematic or not and I also don’t know how much complexity it would involve implementing this. I think I will implement (vibe code) this in a custom rule for now. Maybe I can collect some feedback over time and see if it catches atucal problems. If so, I will report back and maybe it can be considered again.

Once more: thank you for the detailed feedback.
